code/backend/batch/utilities/tools/QuestionAnswerTool.py

```diff
@@ -20,31 +30,115 @@ def __init__(self) -> None:
+        sources_text = "\n\n".join(
+            [f"[doc{i+1}]: {source.page_content}" for i, source in enumerate(sources)]
+        )
+
+        return answering_prompt, {
+            "question": question,
+            "sources": sources_text,
+        }
+
+    def generate_llm_chain(
+        self,
+        question: str,
+        chat_history: list[dict],
+        sources: list[Document],
+    ):
+        examples = []
+
+        few_shot_example = {
+            "documents": self.config.example.documents.strip(),
+            "user_question": self.config.example.user_question.strip(),
+            "answer": self.config.example.answer.strip(),
+        }
+
+        # Remove whitespace
+        if few_shot_example["documents"]:
+            few_shot_example["documents"] = json.dumps(
+                json.loads(few_shot_example["documents"])
+            )
+
+        if any(few_shot_example.values()):
+            if all((few_shot_example.values())):
+                examples.append(few_shot_example)
+            else:
+                logger.warning(
+                    "Not all example fields are set in the config. Skipping few-shot example."
+                )
+
+        example_prompt = ChatPromptTemplate.from_messages(
+            [
+                HumanMessagePromptTemplate.from_template(
+                    self.config.prompts.answering_user_prompt
+                ),
+                AIMessagePromptTemplate.from_template("{answer}"),
+            ]
+        )
+
+        few_shot_prompt = FewShotChatMessagePromptTemplate(
+            example_prompt=example_prompt,
+            examples=examples,
+        )
+
+        answering_prompt = ChatPromptTemplate.from_messages(
+            [
+                SystemMessage(content=self.config.prompts.answering_system_prompt),
+                few_shot_prompt,
+                SystemMessage(content=self.env_helper.AZURE_OPENAI_SYSTEM_MESSAGE),
+                MessagesPlaceholder("chat_history"),
+                HumanMessagePromptTemplate.from_template(
+                    self.config.prompts.answering_user_prompt
+                ),
+            ]
+        )
 
+        documents = json.dumps(
+            {
+                "retrieved_documents": [
+                    {f"[doc{i+1}]": {"content": source.page_content}}
+                    for i, source in enumerate(sources)
+                ],
+            }
+        )
+
+        return answering_prompt, {
+            "user_question": question,
+            "documents": documents,
+            "chat_history": chat_history,
+        }
+
+    def answer_question(self, question: str, chat_history: list[dict], **kwargs: dict):
         # Retrieve documents as sources
         sources = self.vector_store.similarity_search(
             query=question, k=4, search_type="hybrid"
         )
 
-        # Generate answer from sources
+        # If answering_prompt has been set, then use legacy_generate_llm_chain for backwards compatibility
```

Comment:
Should we log out a warning if this is being used with a nudge to migrate? I'm guessing we'll want to remove this in future.

Reply:
Good idea, added a warning and configured logging